### PR TITLE
Experimental Editing Commands

### PR DESCRIPTION
**Pull request by ggarra13**
This commit is for:

Implements https://github.com/AcademySoftwareFoundation/OpenTimelineIO/issues/1649.
Makes PR https://github.com/AcademySoftwareFoundation/OpenTimelineIO/pull/1518,  PR #1650 and PR #1651 obsolete.

It implements edit commands in the otio::algo:: namespace. The commands implemented are those discussed in https://github.com/AcademySoftwareFoundation/OpenTimelineIO/issues/711.

They are further explained more clearly in:
https://github.com/mccartnm/OpenTimelineIO/blob/openedit_design/docs/design/editorial_design.md

Summarize your change.

It implements the needed base commands for editing with otio, building on top of @darbyjohnston's PR.

@darbyjohnston 's PR https://github.com/AcademySoftwareFoundation/OpenTimelineIO/pull/1518 was the base for it, but I debugged two subtle precision bugs in his code and moved it all to the algo:: namespace.

Implementation also modifies errorStatus.h/cpp to add NOT_A_GAP enum.

Implementation, as @darbyjohnston's, makes Composition's protected _index_of_child a now public member of the API as index_of_child and updates track.cpp accordingly.

Reference associated tests.

test_editAlgorithm.cpp is added as in @darbyjohnston, but with tests for all commands.    

I have now tested the commands on an actual application.  The commands overwrite, insert, slice have been tested on mrv2.

Signed-off-by: Gonzalo Garramuño <ggarra13@gmail.com>

### CODE REVIEW

File: src/opentimelineio/algo/editAlgorithm.h [hunk 1]
```diff
@@ -0,0 +1,188 @@
...
+    Fit
+};
+
+// Overwrite an item or items.
+//
+// | A | B |  ->  |A| C |B|
+//   ^   ^
+//   | C |
+//
+//          item = item to overwrite (usually a clip)
```
Comment (meshula) [inline src/opentimelineio/algo/editAlgorithm.h:24]: I think this is C in the diagram, could you make a note?

Could you write some notes about the corner cases?

C starts or ends before A
C starts or ends after B

File: src/opentimelineio/algo/editAlgorithm.h [hunk 2]
```diff
@@ -0,0 +1,188 @@
...
+    bool             remove_transitions = true,
+    Item*            fill_template      = nullptr,
+    ErrorStatus*     error_status       = nullptr);
+
+// Insert an item.
+// |     A     | B |  ->  | A | C | A | B |
+//       ^
+//     | C |
+//
+//          item = item to insert (usually a clip)
```
Comment (meshula) [inline src/opentimelineio/algo/editAlgorithm.h:43]: I assume that if length of AB is L1, and length of C is L2, then length of the result is L1+L2? In other words A is split?

File: src/opentimelineio/algo/editAlgorithm.h [hunk 3]
```diff
@@ -0,0 +1,188 @@
...
+    bool const          remove_transitions = true,
+    Item*               fill_template      = nullptr,
+    ErrorStatus*        error_status       = nullptr);
+
+//
+// Adjust a single item's start time or duration.
+//
+// |    A    | B | C |  ->  |  A  |FILL| B | C |
+//        <--*
+//
```
Comment (meshula) [inline src/opentimelineio/algo/editAlgorithm.h:63]: If FILL is nullptr, then is the result of the trim

| A | B | C |

in other words,  B and C slip to the left?

File: src/opentimelineio/algo/editAlgorithm.h [hunk 4]
```diff
@@ -0,0 +1,188 @@
...
+    RationalTime const& time,
+    bool const          remove_transitions = true,
+    ErrorStatus*        error_status       = nullptr);
+
+//
+// Slip an item start_time by + or -, clamping to available_range if available.
+//
+// |   A   |
+//  <----->
+//
```
Comment (meshula) [inline src/opentimelineio/algo/editAlgorithm.h:95]: how is the slip accomplished? what happens if the delta results in negative values?

File: src/opentimelineio/algo/editAlgorithm.h [hunk 5]
```diff
@@ -0,0 +1,188 @@
...
+//          item = item to slip (usually a clip)
+//         delta = +/- rational time to slip the item by.
+void slip(Item* item, RationalTime const& delta);
+
+//
+// Slide an item start_time by + or -, moving the previous item's duration.
```
Comment (meshula) [inline src/opentimelineio/algo/editAlgorithm.h:101]:
```suggestion
// Slide an item start_time by + or -, adjusting the previous item's duration.
```

File: src/opentimelineio/algo/editAlgorithm.h [hunk 6]
```diff
@@ -0,0 +1,188 @@
...
+// Clamps previous item's duration to available_range if available.
+//
```
Comment (meshula) [inline src/opentimelineio/algo/editAlgorithm.h:103]: what if there is no previous item?

File: src/opentimelineio/algo/editAlgorithm.h [hunk 7]
```diff
@@ -0,0 +1,188 @@
...
+// Adjust a source_range without affecting any other items.
+//
+// |   A   |   B   |  ->  | A |  B  |FILL|
+//      <--*
+//            
+//      item = Item to apply ripple to (usually a clip)
+//  delta_in = RationalTime that the item's source_range().start_time()
+//             will be adjusted by
+// delta_out = RationalTime that the item's
+//             source_range().end_time_exclusive() will be adjusted by
```
Comment (meshula) [inline src/opentimelineio/algo/editAlgorithm.h:121]: what happens if the adjustment results in an element with zero or negative duration?

File: src/opentimelineio/algo/editAlgorithm.h [hunk 8]
```diff
@@ -0,0 +1,188 @@
...
+// End time in parent of Item after input item will never change
+//
+// |   A   |   B   |  ->  | A |  B      |
+//      <--*
+//            
+//      item = Item to apply roll to (usually a clip)
+//  delta_in = RationalTime that the item's source_range().start_time()
+//             will be adjusted by
+// delta_out = RationalTime that the item's
+//             source_range().end_time_exclusive() will be adjusted by
```
Comment (meshula) [inline src/opentimelineio/algo/editAlgorithm.h:143]: same question if result is a zero or negative length item?

File: src/opentimelineio/algo/editAlgorithm.h [hunk 9]
```diff
@@ -0,0 +1,188 @@
...
+//            
+// | A | C | B |  ->  | A |GAP| B |
+//       ^
+//       |
+//
+//           track = track to remove item from.
+//            time = RationalTime
+//            fill = whether to fill the hole with fill_template.
+//   fill_template = if nullptr, use a gap to fill the hole.
+//
```
Comment (meshula) [inline src/opentimelineio/algo/editAlgorithm.h:180]: If GAP is not supplied, does this become a ripple?

**Review by meshula — COMMENTED** [review]
This is fantastic. I have some questions that I am hoping you can answer as new comment documentation, where I have asked the questions? 

Someone can discover the answers to the questions by reading the code, but I think it would be good to have the explanation in the header for easy reference.

File: src/opentimelineio/algo/editAlgorithm.h (hunk 3 above)
Comment (ggarra13) [reply src/opentimelineio/algo/editAlgorithm.h:78]: No.  It always creates a gap.  Clarified in new comments.

File: src/opentimelineio/algo/editAlgorithm.h (hunk 4 above)
Comment (ggarra13) [reply src/opentimelineio/algo/editAlgorithm.h:115]: They are accepted as long as there's no available range or the available range supports the negative values.

File: src/opentimelineio/algo/editAlgorithm.h (hunk 5 above)
Comment (ggarra13) [reply src/opentimelineio/algo/editAlgorithm.h]: Changed.

File: src/opentimelineio/algo/editAlgorithm.h (hunk 7 above)
Comment (ggarra13) [reply src/opentimelineio/algo/editAlgorithm.h:148]: It is clamped.

**Review by meshula — APPROVED** [review]
thanks for updating the docs

**Pull request merged**